geotrellis/geotrellis-site
# Add terraform website resources

# Overview

Previously, the Geotrellis website was built/deployed from the [geotrellis-site-deployment](https://github.com/geotrellis/geotrellis-site-deployment) repo using git submodules. As part of an ongoing effort to deploy each service individually, this PR adds terraform resources for a geotrellis-site ECS service.

# Changes
- [ ] Replace Makefile with `scripts/infra`
- [ ] Use the [`terraform-aws-ecs-web-service`](https://github.com/azavea/terraform-aws-ecs-web-service) to create an ECS task for the website.
- [ ] Place a Cloudfront distribution in front of the Website service to cache tiles.
- [ ] Use terraform data sources to access outputs from the core infrastructure remote state.
- [ ] Add DNS and firewall resources relevant to the web service
- [ ] Ensure all nginx assets are being served over HTTPS

# Testing
- Run `scripts/infra plan`, ensure no changes are necessary.
```bash
$ export GIT_COMMIT=586ee3a
$ export GT_SITE_SETTINGS_BUCKET=geotrellis-site-production-config-us-east-1
$ ./scripts/infra plan
```

- See [https://preview.geotrellis.io](https://preview.geotrellis.io). 

# Notes
- The `gt-transit` and `gt-chatta-demo` services have not been updated yet, so those aspects of the site are unavailable. In the task definition, I set the `extraHosts` for those two services to redirect to `127.0.0.1`. Ultimately, `gt-chatta-demo` and `gt-transit` will have their own services with DNS names that the website can resolve.

- https://www.preview.geotrellis.io and https://beta.preview.geotrellis.io aren't working because there's a pending request for an SSL certificate. Once that's approved, `ssl_certificate_arn` will need to be updated.

- I added `aws_route_53_records` for the `beta`, `www`, and naked `preview.geotrellis.io` subdomains so that the DNS is codified. However, this creates a circular condition where the Route 53 resources require CloudFront to exist, but CloudFront can't create a CDN if Route 53 records for the Alternate CNAMEs don't exist. I skirted this issue by creating the CloudFront distribution first (with no associated CNAMEs), Creating the Route 53 records, and then adding the CNAMEs afterwards. This issue only presents itself when deploying into an empty environment, but it may be worth keeping the Route 53 records outside of terraform control. 

Requires geotrellis/geotrellis-site-deployment#15
Fixes #55 and #61
